File: backoff/res/scripts/zombie_searchingstate.lua
mapManager = luajava.bindClass( "com.deadtroll.backoff.engine.manager.MapManager" )
mathUtil = luajava.bindClass( "com.deadtroll.backoff.engine.helper.MathUtil" )
gameScene = luajava.bindClass( "com.deadtroll.backoff.scene.GameScene")

function random_destination(zombie) 
	if (zombie:getDestination() == nil or math.random() < 0.01) then
		randomX = math.random()*gameScene.WORLD_WIDTH
		randomY = math.random()*gameScene.WORLD_HEIGHT
		newDestination = luajava.newInstance("org.newdawn.slick.geom.Vector2f", randomX, randomY)
		return newDestination
	end
	return zombie:getDestination()
end

function act(zombie)	
    destinationDistance = zombie:getCenter():distance(zombie:getDestination())
	if (destinationDistance > 20.0) then
		if (not zombie:isCollided()) then
			zombie:setPreviousRotation(zombie:getRotation())
			zombie:setPreviousPosition(zombie:getPosition())
			viewPort = mapManager:getInstance():getViewPort()
			destination = zombie:getDestination()
		
			vectToDestination = destination:copy():sub(zombie:getPosition())
			heading = vectToDestination:getNormal()
			
			heading:scale(zombie:getSpeed())
			newPosition = zombie:getPosition():add(heading)
			
			zombie:setPosition(newPosition)
			zombie:setRotation(mathUtil:calcAngle2D(zombie:getPosition(), destination))
		else
			zombie:setCollided(false)
			zombie:setRotation(zombie:getPreviousRotation())
			prev_position = zombie:getPreviousPosition()
			if (prev_position == nil) then
				prev_position = zombie:getPosition()
			end
			zombie:setPosition(prev_position)			
		end
	else
		zombie:setDestination(random_destination(zombie))
	end
end

File: trunk/backoff/src/com/deadtroll/backoff/model/enemy/ZombieSearchingState.java
package com.deadtroll.backoff.model.enemy;

import org.newdawn.slick.geom.Vector2f;

import com.deadtroll.backoff.engine.ai.fsm.State;
import com.deadtroll.backoff.engine.manager.EntityManager;
import com.deadtroll.backoff.engine.model.player.IPlayer;
import com.deadtroll.backoff.engine.resource.ResourceLoader;
import com.deadtroll.backoff.engine.resource.scripting.LuaScript;
import com.deadtroll.backoff.scene.GameScene;

public class ZombieSearchingState extends State {
	private StateMachineZombie zombie;
	private LuaScript luaScript;	

	public ZombieSearchingState(StateMachineZombie zombie) {
		this.zombie = zombie;
		this.zombie.setDestination(randomDestination());
		this.luaScript = ResourceLoader.getInstance().getResource("zombieSearchingState");
	}
	
	private Vector2f randomDestination() {
		if (this.zombie.getDestination() == null || Math.random() < 0.01) {
			return new Vector2f((float)Math.random()*GameScene.WORLD_WIDTH, (float)Math.random()*GameScene.WORLD_HEIGHT);
		}
		return this.zombie.getDestination();
	}

	@Override
	public void act() {
		this.luaScript.doCall("act", this.zombie);
	}

	@Override
	public State getNextState() {
		IPlayer player = EntityManager.getInstance().getPlayer();
		float playerDistance = player.getCenter().distance(this.zombie.getCenter());
		float destinationDistance = this.zombie.getCenter().distance(this.zombie.getDestination());
		
		if (playerDistance > 100.0) {
			if (destinationDistance < 20.0f)
				return new ZombieDestinationReachedState(this.zombie);
			else
				return this;
		} else {
			return new ZombieChasingState(this.zombie);
		}
	}
}
